### PR TITLE
[Clang][AArch64] Generalise streaming mode checks for builtins.

### PR DESCRIPTION
**Pull request by sdesmalen-arm**
PR #76975 added 'IsStreamingOrSVE2p1' to emit a diagnostic when a builtin marked
with 'IsStreamingOrSVE2p1' is used in a non-streaming function that is not
compiled with `+sve2p1`.

The problem is a bit more complex than only this case. For example, we've marked
lots of builtins with 'IsStreamingCompatible', meaning it can be used in either
streaming, streaming-compatible or non-streaming functions. But the code in
SemaChecking, doesn't check the appropriate target guards. This issue becomes
relevant when SVE builtins are only available in streaming mode, e.g. when
compiling for SME without SVE.

If we were to add the appropriate target guards, we'd have to add many more
combinations, e.g.:
  IsStreamingSMEOrSVE
  IsStreamingSME2OrSVE2
  IsStreamingSMEOrSVE2p1
  IsStreamingSME2OrSVE2p1
  etc.

To avoid having to add more combinations (and avoid having to add more in the
future for new extensions), we use a single 'IsSVEOrStreamingSVE' flag for all
builtins that are available in streaming mode for the appropriate SME flags, or
in non-streaming mode for the appropriate SVE flags, or both.  The code in
SemaChecking will then verify for which mode (or both) the builtin would be
defined, given the target features of the function/compilation unit.

For example:

  'svclamp' is enabled under FEAT_SVE2p1 and FEAT_SME2

* When we compile for SVE2p1 and SME (but not SME2), the builtin is undefined
  behaviour when called from a streaming function.

* When we compile for SME2 and SVE2 (but not SVE2p1), the builtin is undefined
  behaviour when called from a non-streaming function.

* When we compile for _both_ SVE2p1 and SME2, the builtin can be used in either
  mode (non-streaming, streaming or streaming-compatible)

### CODE REVIEW

File: clang/test/Sema/aarch64-streaming-sme-or-nonstreaming-sve-builtins.c [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+__attribute__((target("+sme2")))
+svfloat32_t good2(svfloat32_t a, svfloat32_t b, svfloat32_t c) __arm_streaming {
+  return svclamp(a, b, c);
+}
+
+__attribute__((target("+sve2p1,+sme2")))
+svfloat32_t good3(svfloat32_t a, svfloat32_t b, svfloat32_t c) __arm_streaming_compatible {
+  return svclamp(a, b, c);
+}
```
Comment (aemerson) [inline clang/test/Sema/aarch64-streaming-sme-or-nonstreaming-sve-builtins.c:21]: For completeness can we have a test as well to check this case but instead with `__arm_streaming` as well

**Review by aemerson — APPROVED** [review]
LGTM with suggestion

File: clang/lib/Sema/SemaARM.cpp [hunk 2]
```diff
@@ -561,16 +561,61 @@ SemaARM::ArmStreamingType getArmStreamingFnType(const FunctionDecl *FD) {
...
+
+      S.Context.getFunctionFeatureMap(CallerFeatureMap, FD);
+      DisableFeatures("sve");
+      SatisfiesSME = Builtin::evaluateRequiredTargetFeatures(
+          BuiltinTargetGuards, CallerFeatureMap);
+
+      CachedFD = FD;
+    }
+
+    if (SatisfiesSVE && SatisfiesSME)
```
Comment (paulwalker-arm) [inline clang/lib/Sema/SemaARM.cpp:610]: Does this effectively prevent streaming compatible functions when only SVE feature flags are available?

My updated understand of streaming compatible functions is that SME features play no role and the user is expected to use SVE feature flags to direct the compiler to the level of SVE support a streaming compatible function can have, much like they would for ordinary functions.

File: clang/include/clang/Basic/arm_sve_sme_incl.td [hunk 3]
```diff
@@ -225,7 +225,7 @@ def IsStreamingCompatible           : FlagType<0x4000000000>;
 def IsReadZA                        : FlagType<0x8000000000>;
 def IsWriteZA                       : FlagType<0x10000000000>;
 def IsReductionQV                   : FlagType<0x20000000000>;
-def IsStreamingOrSVE2p1             : FlagType<0x40000000000>; // Use for intrinsics that are common between sme/sme2 and sve2p1.
+def IsSVEOrStreamingSVE             : FlagType<0x40000000000>; // Use for intrinsics that are common between SVE and SME.
```
Comment (paulwalker-arm) [inline clang/include/clang/Basic/arm_sve_sme_incl.td:228]: If you permit a bit of bike shedding I don't think this is a good name.  From what I can see the new flag is used to trigger dynamic resolution to determine if the builtin is available to use based on the target features along with any keywords associated with the function.  Perhaps `RequiresDynamicVerification`? or ideally something shorter that has the same meaning.

**Review by paulwalker-arm — CHANGES_REQUESTED** [review]

File: clang/lib/Sema/SemaARM.cpp [hunk 4]
```diff
@@ -559,31 +559,86 @@ SemaARM::ArmStreamingType getArmStreamingFnType(const FunctionDecl *FD) {
...
-      BuiltinType = SemaARM::ArmStreamingCompatible;
-    else
+
+  // Check if the intrinsic is available in the right mode, i.e.
+  // * When compiling for SME only, the caller must be in streaming mode.
+  // * When compiling for SVE only, the caller must be in non-streaming mode.
+  // * When compiling for both SVE and SME, the caller can be in either mode.
+  if (BuiltinType == SemaARM::VerifyRuntimeMode) {
+    static llvm::StringMap<bool> CallerFeatureMapWithoutSVE,
+        CallerFeatureMapWithoutSME;
```
[the remark this answers is not in the document]
Comment (sdesmalen-arm) [reply clang/lib/Sema/SemaARM.cpp]: You're right, not sure what I was thinking here with the 'static' variable :)

I'd rather not go down the route of parsing the BuiltinTargetGuards here or making assumptions on the format if we're not going to ensure this format in the arm_sve.td file. For example, one can write `(sve,featureX)|(sme2,featureX)` or `(sve|sme),featureX`. Sure we can assume the former format, but in that case I think we need to have a `SVETargetGuard` and a `SMETargetGuard`, that we let TableGen combine into a canonical form for the combined `TargetGuard`.

File: clang/lib/Sema/SemaARM.cpp [hunk 5]
```diff
@@ -622,7 +679,8 @@ bool SemaARM::CheckSMEBuiltinFunctionCall(unsigned BuiltinID,
     }
 
     if (BuiltinType)
-      checkArmStreamingBuiltin(SemaRef, TheCall, FD, *BuiltinType);
+      HasError |= checkArmStreamingBuiltin(SemaRef, TheCall, FD, *BuiltinType,
```
[the remark this answers is not in the document]
Comment (sdesmalen-arm) [reply clang/lib/Sema/SemaARM.cpp]: Done

File: clang/lib/Sema/SemaARM.cpp [hunk 6]
```diff
@@ -559,31 +559,86 @@ SemaARM::ArmStreamingType getArmStreamingFnType(const FunctionDecl *FD) {
...
+  // * When compiling for SME only, the caller must be in streaming mode.
+  // * When compiling for SVE only, the caller must be in non-streaming mode.
+  // * When compiling for both SVE and SME, the caller can be in either mode.
+  if (BuiltinType == SemaARM::VerifyRuntimeMode) {
+    static llvm::StringMap<bool> CallerFeatureMapWithoutSVE,
+        CallerFeatureMapWithoutSME;
+
+    // Cache the feature maps, to avoid having to recalculate this for each
+    // builtin call.
+    static unsigned CachedODRHash = 0;
```
[the remark this answers is not in the document]
Comment (sdesmalen-arm) [reply clang/lib/Sema/SemaARM.cpp]: Done

**Review by paulwalker-arm — APPROVED** [review]

**Pull request merged**